### PR TITLE
refactor(copy): calc suitable parallelism for reading parquet splits.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1.  columns of a  row group have to be all read before processing, so read as fast as possible in parallel (for simplicity,  not consider max_io_request here). 
2. trait num of fields F as an io unit.,  we calc parallel in splits N , which limited by 
   1. sum(largest N split) < max_memory 
   3. N < max_io_request/F

note:

1. it not work for `load with transform`  and stage table function
2. current default  max_storage_io_requests = 64, may be too small for this? e.g. 

```
  2023-03-21T11:34:13.315663Z  INFO common_pipeline_sources::input_formats::input_pipeline: copy read 1 splits in parallel, according to max_memory=27487790694, num_fields=109, max_storage_io_requests=48, max_split_size=107581

```

to do (not this pr)

-  a row group is read now with parallelism = "num of top fields" in this pr, it should be num of leaf columns.
-  there may still be blocks in ports between processors and in processing,  DeserializeSource should try to cut large blocks into smaller ones.    before that,  user can set smaller max_threads

Closes #issue

### CODE REVIEW

File: src/query/pipeline/sources/src/input_formats/impls/input_format_parquet.rs [hunk 1]
```diff
@@ -489,3 +506,27 @@ pub fn split_column_metas_by_field(
...
+    columns: &'a [ColumnChunkMetaData],
+    field_name: &str,
+) -> Vec<&'a ColumnChunkMetaData> {
+    columns
+        .iter()
+        .filter(|x| x.descriptor().path_in_schema[0] == field_name)
+        .collect()
+}
+
+async fn _read_single_column_async<R>(
```
Comment (ariesdevil) [inline src/query/pipeline/sources/src/input_formats/impls/input_format_parquet.rs:520]: Why does this function's name starts with underline?

Comment (youngsofun) [reply src/query/pipeline/sources/src/input_formats/impls/input_format_parquet.rs]: done

File: src/query/pipeline/sources/src/input_formats/input_pipeline.rs [hunk 2]
```diff
@@ -191,39 +189,59 @@ pub trait InputFormatPipe: Sized + Send + 'static {
...
-                    while let Some(row_batch) = futs.next().await {
-                        match row_batch {
-                            Ok(row_batch) => {
-                                if row_batch_tx.send(row_batch).await.is_err() {
-                                    break;
-                                }
-                            }
-                            Err(cause) => {
-                                row_batch_tx.send(Err(cause)).await.ok();
+        let max_storage_io_requests = ctx.settings.get_max_storage_io_requests().unwrap();
```
Comment (BohuTANG) [inline src/query/pipeline/sources/src/input_formats/input_pipeline.rs:194]:
```suggestion
        let max_storage_io_requests = ctx.settings.get_max_storage_io_requests()?;
```

File: src/query/pipeline/sources/src/input_formats/input_pipeline.rs [hunk 3]
```diff
@@ -191,39 +189,59 @@ pub trait InputFormatPipe: Sized + Send + 'static {
...
+        let per_split_io = ctx.schema.fields().len();
+        let max_splits = max_storage_io_requests as usize / per_split_io;
+        let mut max_splits = std::cmp::max(max_splits, 1);
+        let mut sizes = ctx.splits.iter().map(|s| s.size).collect::<Vec<usize>>();
+        sizes.sort_by(|a, b| b.cmp(a));
+        let max_memory = ctx.settings.get_max_memory_usage().unwrap() as usize;
```
Comment (BohuTANG) [inline src/query/pipeline/sources/src/input_formats/input_pipeline.rs:200]:
```suggestion
        let max_memory = ctx.settings.get_max_memory_usage()? as usize;
```

**Review by BohuTANG — APPROVED** [review]

**Comment by youngsofun** [comment]
```
error[E0432]: unresolved import `crate::plans::Unnest`
  --> src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:37:5
   |
37 | use crate::plans::Unnest;
   |     ^^^^^^^^^^^^^^^^^^^^ no `Unnest` in `planner::plans`

error[E0599]: no variant or associated item named `Unnest` found for enum `planner::plans::scalar_expr::ScalarExpr` in the current scope
   --> src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:252:29
    |
252 |                 ScalarExpr::Unnest(unnest) => {
    |                             ^^^^^^ variant or associated item not found in `ScalarExpr`
    |
   ::: src/query/sql/src/planner/plans/scalar_expr.rs:33:1
    |
33  | pub enum ScalarExpr {
    | ------------------- variant or associated item `Unnest` not found for this enum

error[E0599]: no variant or associated item named `Unnest` found for enum `planner::plans::scalar_expr::ScalarExpr` in the current scope
   --> src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:259:36
    |
259 |                     Ok(ScalarExpr::Unnest(Unnest {
    |                                    ^^^^^^ variant or associated item not found in `ScalarExpr`
    |
   ::: src/query/sql/src/planner/plans/scalar_expr.rs:33:1
    |
33  | pub enum ScalarExpr {
    | ------------------- variant or associated item `Unnest` not found for this enum
```

seems caused by https://github.com/datafuselabs/databend/pull/10683 
cc @leiysky  @BohuTANG

**Comment by BohuTANG** [comment]
@mergify update

**Pull request merged**